src/parthenon_arrays.hpp

```diff
@@ -0,0 +1,315 @@
+  }
+
+  #define SLC0 std::make_pair(0,1)
+  KOKKOS_INLINE_FUNCTION
+  auto SliceD(index_pair_t slc, std::integral_constant<int,6>) const {
+    return Slice(slc,Kokkos::ALL(),Kokkos::ALL(),
+                 Kokkos::ALL(),Kokkos::ALL(),Kokkos::ALL());
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto SliceD(index_pair_t slc, std::integral_constant<int,5>) const {
+    return Slice(SLC0,slc,Kokkos::ALL(),
+                 Kokkos::ALL(),Kokkos::ALL(),Kokkos::ALL());
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto SliceD(index_pair_t slc, std::integral_constant<int,4>) const {
+    return Slice(SLC0,SLC0,slc,
+                 Kokkos::ALL(),Kokkos::ALL(),Kokkos::ALL());
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto SliceD(index_pair_t slc, std::integral_constant<int,3>) const {
+    return Slice(SLC0,SLC0,SLC0,
+                 slc,Kokkos::ALL(),Kokkos::ALL());
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto SliceD(index_pair_t slc, std::integral_constant<int,2>) const {
+    return Slice(SLC0,SLC0,SLC0,
+                 Kokkos::ALL(),slc,Kokkos::ALL());
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto SliceD(index_pair_t slc, std::integral_constant<int,1>) const {
+    return Slice(SLC0,SLC0,SLC0,Kokkos::ALL(),Kokkos::ALL(),slc);
+  }
+  #undef SLC0
+
+  // AthenaArray.InitWithShallowSlice(src,dim,indx,nvar)
+  // translates into auto dest = src.SliceD<dim>(std::make_pair(indx,indx+nvar))
+  template<std::size_t N = 6>
+  auto SliceD(index_pair_t slc,
+              std::integral_constant<int,N> ic =
+              std::integral_constant<int,N>{}) const {
+    return SliceD(slc, ic);
+  }
+
+  // AthenaArray.InitWithShallowSlice(src,dim,indx,nvar)
+  // translates into auto dest = src.SliceD<dim>(indx,nvar)
+  template<std::size_t N = 6>
+  auto SliceD(const int indx, const int nvar,
+              std::integral_constant<int,N> ic =
+              std::integral_constant<int,N>{}) {
+    return SliceD(std::make_pair(indx,indx+nvar),ic);
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto Get(int i) const {
+    return Kokkos::subview(d6d_,i,
+                           Kokkos::ALL(),Kokkos::ALL(),
+                           Kokkos::ALL(),Kokkos::ALL(),Kokkos::ALL());
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto Get(int j, int i) const {
+    return Kokkos::subview(d6d_,j,i,
+                           Kokkos::ALL(),Kokkos::ALL(),
+                           Kokkos::ALL(),Kokkos::ALL());
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto Get(int k, int j, int i) const {
+    return Kokkos::subview(d6d_,k,j,i,
+                           Kokkos::ALL(),Kokkos::ALL(),Kokkos::ALL());
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto Get(int n, int k, int j, int i) const {
+    return Kokkos::subview(d6d_,n,k,j,i,
+                           Kokkos::ALL(),Kokkos::ALL());
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto Get(int m, int n, int k, int j, int i) const {
+    return Kokkos::subview(d6d_,m,n,k,j,i,
+                           Kokkos::ALL());
+  }
+  KOKKOS_INLINE_FUNCTION
+  auto Get(int l, int m, int n, int k, int j, int i) const {
```

Comment:
This is just being nitpicky as I don't really know that much about kokkos, is there any reason why the indices can't all be made const?

Reply:
Those can be constants, and that's an improvement. Nice catch.

Reply:
I tend to prefer never marking value-type function parameters as const because it alters the signature of the function without actually impacting the caller in anyway.

Reply:
But it's a small thing.

Reply:
@AndrewGaspar your comment got me thinking and so I googled it. Apparently this is surprisingly controversial! People seem pretty evenly split about which way is right. My instinct leans towards yours in non-critical bits and to pepper `const` everywhere I can in performance-critical bits of code in the hopes it helps the compiler optimize. It's probably superstition---the compiler probably optimizes exactly the same.

Reply:
`const` doesn't improve optimizability of your code since C++ has non-existent guarantees about const data. Anywhere that the compiler is able to guarantee the constness of a value is respected, it can also guarantee it for non-const data. It's more a maintenance feature. In other languages where "const"ness has stronger guarantees (like Rust), it could be theoretically used by the compiler to make optimization decisions.

Reply:
Confirmed: what I was doing is superstition.

Reply:
I will say that while const might not tell the compiler much, it can tell future developers something and the compiler will tell you when you are breaking the rules. So I think it can still communicate intent and that can potentially be useful.